### DISCUSSION TITLE
Support for reliable sessions

### DISCUSSION THREAD

**Discussion by geketi** [CoreWCF/CoreWCF#793] [Q&A]
I am looking to port a project from .net framework to .net 6. The blocking feature that is currently missing from corewcf appears to be reliable sessions. From what I have read this requires non-trivial implementation effort, but is high on the priority list. 

Will there be a chance of this feature releasing this year?

**Comment by geketi** [comment]
Is there room for contribution on this feature? Lack of support for reliable sessions is blocking my organization from moving our IoT solution to .net 6.

**Reply by mconnew** [reply]
I'm going to be doing this soon. You can if you like, but the code is ridiculously complicated in .NET Framework and has a pretty significant bug which limits the rate of connection establishment that I'm going to fix at the same time. The feature ties into the code for TransportWithMessageCredentials feature as it uses a session disconnected from the actual transport session and a lot of that code has been refactored. If you want to have a go at it, let me know. Otherwise what's on my plate coming up is NamedPipe support (client and server) and ReliableSessions so it should be very soon, within the next few months.

**Reply by geketi** [reply]
I've looked at the framework implementation and yes, it is complex to the point I don't think I would be able to contribute. Reliable Sessions being available in a few months would align with our planning.

Thank you for all your work on this project Matt!

**Reply by mconnew** [reply]
One thing which will be a massive help is in testing and verifying the implementation in a real world scenario.

**Reply by topsterde** [reply]
I can also run real-world tests, when NetTcpBinding with Reliable Sessions is available.

**Reply by geketi** [reply]
@mconnew 
@topsterde 

Matthew, hope you are well. Does it still look like support for NetTcpBinding  with Reliable Sessions is in the cards for this year? As soon as you have any code ready for testing, I am ready to immediately provide feedback. The need to migrate our Edge device to .net 6 is getting acute. Thanks.

**Reply by mconnew** [reply]
Still aiming for it. I hit some road bumps that I've had to resolve on my NetNamedPipe implementation on the client (had to refactor all of our pacakges) which has taken a lot longer than I had hoped. After that's complete, I'll implement NamedPipe on CoreWCF which will be a lot quicker as the same issues aren't there and I planned ahead to enable it there. ReliableSessions will be next after that.

**Reply by chuckgmthompson** [reply]
any update on this as we are in need of reliable sessions for the tcpBinding as well?

**Reply by geketi** [reply]
@mconnew
Hi Matthew, saw in the other thread you were sick, sorry to hear that. 

Just need to ask: will reliable sessions be part of the preview release? 

Also, why is a wcf client part of this project, does .net 6.0 not ship with one already? Is the thought that Microsoft might deprecate the client as well?

**Reply by mconnew** [reply]
The wcf client isn't part of this project, but I work on both and was giving some context to the delays. The WCF client doesn't ship with .NET Core, but is a separate nuget package. There's zero plans to deprecate the client, every step towards supporting more of WCF has come as a result of the continued high demand for it, which BTW is still growing. Here's an interesting nuget trends chart comparing the WCF client package downloads with the gRPC client downloads. In the last year WCF has had 80 million downloads, and the gRPC client has had 35 million. You can see the slope is steeper for WCF so that gap is still increasing. It's not going anyway anytime soon. These numbers can't be used to extrapolate number of people using it, but they can be used in a qualitative way to show there's still significant growth.  
https://nugettrends.com/packages?months=12&ids=System.ServiceModel.Primitives&ids=Grpc.Net.Client

**Reply by geketi** [reply]
@mconnew  
Matt, 

About reliable sessions itself, what do you think is a realistic date for us to start testing the feature?

**Reply by pmize3** [reply]
@mconnew
Hi Matt,

Just checking in to see if there are any updates on this?

**Reply by geketi** [reply]
@mconnew 

Hi Matt, do you have any updates on the roadmap for this hotly anticipated feature?

Thanks!

**Reply by mconnew** [reply]
I'm starting back up on ReliableSessions next week.

**Reply by Gena-V** [reply]
Hey @mconnew, any updates on this issue?
Thanks

**Reply by mconnew** [reply]
I have the Request/Reply pattern using WSHttpBinding superficially working (happy case looks to be working) in my branch at https://github.com/mconnew/CoreWCF/tree/ReliableSessions . I tried to use NetTcp and it broke pretty quickly due to some internal bookkeeping not being done correctly. I have a strong suspicion that will make WSHttpBinding also fail in scenarios which should recover, but I haven't got to porting over the failure scenario test cases from .NET Framework yet. Basically WSHttpBinding works sufficiently to kick the tires, but I would strongly warn against using it in production yet. Once I have NetTcpBinding working with it, I'll have more confidence in the code, but I would still wait until I have all the test cases ported.

**Reply by oskarjf** [reply]
I'm also upgrading a .Net Framework WCF solution to CoreWCF where I need bindings with reliable sessions so looking forward to a solution to this :)

**Reply by oskarjf** [reply]
Hi @mconnew . Any ETA for this issue?

**Reply by Moreno-Gentili** [reply]
@mconnew I watched your [video from the .NET Conf 2023](https://youtu.be/zVOAkBMnH5g?si=ksroD4-J3ZUZB3G4&t=1458). Thanks for the update, would you clarify what's left to be done with Reliable Sessions, in case anyone wanted to contribute?

**Reply by geketi** [reply]
@mconnew Any word on this?

**Reply by j-bernhard** [reply]
Is there any update on reliable session. I have a separate Client (.net8 servicemodel) and Server (CoreWCF) implmementation, and don't know how to setup reliable session, is not availabe on CoreWCF (see screenshot). 

![image](https://github.com/CoreWCF/CoreWCF/assets/8314146/ec040e1b-1d67-45f0-8659-fe82da7b0127)

Any hints? Is it supported now, because was mentioned in the .NET Conf video, as upcoming next. 

Thx,
Joerg

**Reply by j-bernhard** [reply]
@mconnew potential any update advice how to setup a reliable session on server side? Is something possible on top by the asp.net middleware?

**Reply by cp-1921** [reply]
> I have the Request/Reply pattern using WSHttpBinding superficially working (happy case looks to be working) in my branch at https://github.com/mconnew/CoreWCF/tree/ReliableSessions . I tried to use NetTcp and it broke pretty quickly due to some internal bookkeeping not being done correctly. I have a strong suspicion that will make WSHttpBinding also fail in scenarios which should recover, but I haven't got to porting over the failure scenario test cases from .NET Framework yet. Basically WSHttpBinding works sufficiently to kick the tires, but I would strongly warn against using it in production yet. Once I have NetTcpBinding working with it, I'll have more confidence in the code, but I would still wait until I have all the test cases ported.

Hey @mconnew is there an estimated timeline for when reliable sessions using NetTcpBinding will be ready for production use?

Thanks

**Comment by cobalto** [comment]
Any news regarding that topic? 😐

**Comment by icnocop** [comment]
This feature is on the Feature Roadmap Vote discussion here:
https://github.com/CoreWCF/CoreWCF/issues/234#issuecomment-703430347

Please vote and thank you.

**Comment by rcocks-hl** [comment]
On the roadmap, this was marked as "In Progress" by @mconnew in November 2023. Is that still the case please?

**Comment by icnocop** [comment]
See PR https://github.com/CoreWCF/CoreWCF/pull/1702
